src/guides/metamask-snap/index.md

```diff
@@ -0,0 +1,1318 @@
+
+  const returnObject: Record<string, unknown> = {
+    message: 'Unknown transaction',
+  };
+  const networkId = await wallet.request({ method: 'net_version' });
+  const SimpleNFTContractAddress = SimpleNft.networks[networkId]? SimpleNft.networks[networkId].address : null;
+  const NFTVaultContractAddress = NFTVault.networks[networkId]? NFTVault.networks[networkId].address : null;
+
+  try {
+    // Check if the transaction has data.
+    if (
+      !isObject(transaction) ||
+      !hasProperty(transaction, 'data') ||
+      typeof transaction.data !== 'string'
+    ) {
+      throw "Transaction data received is not an object."; 
+    }
+
+    switch(transaction.to) { 
+      case SimpleNFTContractAddress.toLowerCase(): 
+        returnObject.message = "You are interacting with the SimpleNFT.sol contract"; 
+        break; 
+      case NFTVaultContractAddress.toLowerCase(): 
+        returnObject.message = "You are interacting with the NFTVault.sol contract"; 
+        break; 
+      default: 
+        returnObject.message = "I do not recognize the address " + transaction.to; 
+    }
+
+    return returnObject;
+  } catch (error) {
+    console.error(error);
+    return returnObject; 
+  }
+}
+```
+
+In this code, we check the `transaction.to` property to see what the calling contract is. If you want to see this in action, we'll have to reinstall the snap. Go back to the browser and hit `reconnect`. MetaMask should prompt you to reinstall. After confirming, mint an NFT. In your MetaMask wallet, you should see that the SimpleNFT contract address was recognized!
+
+![interacted with simple nft](./img/simplenft.gif)
+
+In Solidity, the first 4 bytes of an encoded function (which we demonstrated in the `packages/site` portion of this tutorial, identify what the function is called. We can use this fact to translate what functions are actually being called by the `NFTVault` contract. In the `NFTVault` switch case statement, add:
+
+```javascript
+const transactionData = remove0x(transaction.data);
+
+// Get function signature, i.e., the first 4 bytes of the data.
+const functionSignature = transactionData.slice(0, 8);
+
+let matchingFunction = '';  
+
+switch (functionSignature) {
+  case '4e1ca120':
+    matchingFunction = 'approveWithdraw(address,uint256)';
+    break;
+  case '97be5523':
+    matchingFunction = 'depositNFT(address,uint256,address)';
+    break;
+  case 'b537b269':
+    matchingFunction = 'removeApproval(address,uint256)';
+    break;
+  case '6088e93a':
+    matchingFunction = 'withdrawNFT(address,uint256)';
+    break;
+  default:
+    break;
+}
+
+if(matchingFunction.length > 0) { 
+  returnObject.method = matchingFunction; 
+}
+```
+
+After you reconnect your snap, you should see something like this:
+
+![function insight](./img/function_insight.gif)
+
+To get even more granular, let's actually decode the argument data when `withdrawNFT` is called. First, add this function, which will convert our data to strings:
```

Comment:
This function converts certain arguments which are not JSON serializable to strings. I don't know if that applies to all of our data.